src/Features/CSharp/Portable/Snippets/CSharpWhileLoopSnippetProvider.cs

```diff
@@ -0,0 +1,82 @@
+using Microsoft.CodeAnalysis.Indentation;
+using Microsoft.CodeAnalysis.LanguageService;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Snippets;
+using Microsoft.CodeAnalysis.Snippets.SnippetProviders;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.Snippets
+{
+    [ExportSnippetProvider(nameof(ISnippetProvider), LanguageNames.CSharp), Shared]
+    internal class CSharpWhileLoopSnippetProvider : AbstractWhileLoopSnippetProvider
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CSharpWhileLoopSnippetProvider()
+        {
+        }
+
+        protected override int GetTargetCaretPosition(ISyntaxFactsService syntaxFacts, SyntaxNode caretTarget, SourceText sourceText)
+        {
+            var whileStatement = (WhileStatementSyntax)caretTarget;
+            var blockStatement = (BlockSyntax)whileStatement.Statement;
+
+            var triviaSpan = blockStatement.CloseBraceToken.LeadingTrivia.Span;
+            var line = sourceText.Lines.GetLineFromPosition(triviaSpan.Start);
+            // Getting the location at the end of the line before the newline.
+            return line.Span.End;
+        }
+
+        protected override SyntaxNode GetStatementCondition(SyntaxNode node)
+        {
+            var whileStatement = (WhileStatementSyntax)node;
+            return whileStatement.Condition;
+        }
+
+        private static string GetIndentation(Document document, SyntaxNode node, SyntaxFormattingOptions syntaxFormattingOptions, CancellationToken cancellationToken)
+        {
+            var parsedDocument = ParsedDocument.CreateSynchronously(document, cancellationToken);
+            var whileStatementSyntax = (WhileStatementSyntax)node;
+            var openBraceLine = parsedDocument.Text.Lines.GetLineFromPosition(whileStatementSyntax.Statement.SpanStart).LineNumber;
+
+            var indentationOptions = new IndentationOptions(syntaxFormattingOptions);
+            var newLine = indentationOptions.FormattingOptions.NewLine;
+
+            var indentationService = parsedDocument.LanguageServices.GetRequiredService<IIndentationService>();
+            var indentation = indentationService.GetIndentation(parsedDocument, openBraceLine + 1, indentationOptions, cancellationToken);
+
+            // Adding the offset calculated with one tab so that it is indented once past the line containing the opening brace
+            var newIndentation = new IndentationResult(indentation.BasePosition, indentation.Offset + syntaxFormattingOptions.TabSize);
+            return newIndentation.GetIndentationString(parsedDocument.Text, syntaxFormattingOptions.UseTabs, syntaxFormattingOptions.TabSize) + newLine;
+        }
+
+        protected override async Task<Document> AddIndentationToDocumentAsync(Document document, int position, ISyntaxFacts syntaxFacts, CancellationToken cancellationToken)
```

Comment:
The indentation should probably be abstracted out for every block - since most of them are exactly the same. I'll do that in a follow-up PR

Reply:
The vast majority of snippets have such block, so this definitely needs an abstraction. I am only not sure how it will work if semantic snippets become a VB thing as well